### ISSUE TITLE
Support for 256+ threads (x2APIC & AMD Extended APIC ID)

### ISSUE THREAD

**Issue by debrouxl** [memtest86plus/memtest86plus#70]
At the time of this writing, x86_64 computers with 256+ threads already exist: 2S EPYC 77x[23], 8S Xeon Platinum 82xx for select models. Though they'll remain fairly niche, their market share and ease of access will keep increasing, e.g. with the Zen 4 series, which is rumored to contain some 96-core models - i.e. 384 threads in a 2S configuration, as those models are likely to offer SMT=2 like their EPYC Zen 2 and Zen 3 predecessors did - or probably some Sapphire Rapids Xeon models. There are higher-priority features, but the 256 thread limit is going to hit some new users in the near future.

Technically, the solution to the 256 core limit issue of the original 8-bit APIC is available from Nehalem onwards in Intel processors ("x2APIC", CPUID 1 ecx register bit 21 & CPUID B & CPUID 1F preferred to B if available), and more or less differently on AMD processors (at least "Extended APIC space", with two CPUID-based methods depending on the generation).
The 6272 Bulldozer and 6378 Piledriver Opterons have support for AMD Extended APIC space but do not advertise the x2APIC bit in CPUID 1 ecx. AMD processors from Bulldozer onwards are supposed to set CPUID 80000001 ecx bit 22 (topology extensions) and provide the 32-bit extended APIC ID in CPUID 8000001E eax. On the Ryzen 7 PRO 5750GE (Zen 3), the CPUID data advertises both x2APIC and AMD Extended APIC space.

**Comment by x86fr** [comment]
Adding x2APIC support is nice, but there is a real debate here. For a Memory Testing Tool, does using as much core as possible make the software more efficient to detect memory issues? Massive multi-threading is great to use as much raw computing power as available. There is no doubt about this. But considering that a single core of any modern CPU can retire data at a much higher bandwidth than available from its IMC (even a 4 channel DDR5), why do we need to split a work that can be done on a single thread to 64, 256 or 1024 threads?

Modern HCC CPUs uses architecture like ring bus or chiplets that induce heavy latency due to cache coherency, remote penalty and contention issues on interconnects when many cores try to access the same IMC from different points of the CPU. This has been demonstrated by many research like https://tel.archives-ouvertes.fr/tel-01549294/document. Many Intel tools used to validate the DDR5 bus only use a single core to maximize the bandwidth by removing useless inter-cores transactions. It also creates a single hotspot on the die, more prone to failure than a distributed workload. On AMD platform, IIRC, a single thread is used per chiplet to exercise the IMC (but I need to check again as I'm more familiar with Intel tools). 

We're now sure that using as many core as possible on Hybrid CPU is actually counterproductive as threads on Atom-based core slow down the IMC by a large margin. I wonder if by using so many threads (let's say more than 4), we're not adding unnecessary penalties on memory access that can reduce the effectiveness of fast memory error detection.

Answering this question is extremely hard without knowing exactly how the IMC works, but I guess that using 32, 64, 256 or even more threads instead of 4 (and maybe 1) does not add a significant benefit to detect memory errors in a timely manner. At least, I would like to be sure it's not worst.

I'm in contact with skilled Intel Engineers that can probably help about this. I'll ask them and post the answer here (or better, ask them if they can directly post their answers here). Another option would be to create a scientific test protocol with some known-defective memory modules, and check how many errors are detected in the time frame (ie: 1 thread vs 4 threads vs 16 threads)

**Comment by debrouxl** [comment]
You're entirely right about 256+ _simultaneous_ threads not being of much use, or even being counter-productive, to memory error detection, for multiple reasons. The remote access penalties are noticeable with the current NUMA-unaware test code ( #12 ).

However, without support for x2APIC / Extended APIC ID, how can a memory testing tool / bus validation tool ensure that every thread in a 256+-thread system - even one thread at a time - accesses memory as it is supposed to ?

**Comment by 01e3** [comment]
Also, are we sure that a single thread or even multiple threads, but placed more or less randomly (=suboptimally) is more than sufficient to saturate all memory channels? On top of multi-socket machines, growing number of multi-chiplet CPUs, we also have had multi-ring CPUs for some time [1], and high(er) memory channel count CPUs.

An observation: not sure how much the memory benchmark from memtest86+ is representative here for the testing performance itself, but on all my systems it reports values that are significantly lower than the theoretical BW, and also significantly lower than what AIDA64 reports. I would even expect that during the testing, BW is even smaller than what the benchmark shows, and workstations that are able to support hundreds of GBs of memory are not that expensive these days, where the testing matters...

Perhaps the answer is *the proper* number of active threads, from proper NUMA domains, accessing the proper memory regions? Which probably is going to be non-trivial, beyond just the scope of this FR.

[1] https://www.anandtech.com/show/10158/the-intel-xeon-e5-v4-review/2

**Comment by 01e3** [comment]
By a coincidence, I found https://infobazy.gda.pl/2014/pliki/prezentacje/d2s2e4-Kaczmarski-Optymalna.pdf

The most important takeaway:
_There is a limitation on the number of memory transactions that can be issued by each core. When only a few cores are active, there are not enough memory transactions to saturate the memory controller and take advantage of the higher frequency._

**Comment by x86fr** [comment]
In real world applications, each core only access a limited amount of memory, so it might be required to use many cores to saturate the memory controller. The quantity of data a single core can generate on synthetic application like Memtest86+ is much higher than available on any memory controller, virtually unlimited. That said, on multi socket server platforms, using one core per socket might be required to saturate the bandwidth of all memory controller available. It depends on the bandwidth of the interconnection bus used between sockets.

**Comment by 01e3** [comment]
Fair point about the synthetic application but please note that is a huge gap between Memtest86+ benchmark results and what AIDA64 shows, where I would expect the benchmark should be the best-case scenario, and clearly a "synthetic workload" to your point.

For example, here is AIDA64 vs Memtest86 for E5-2667v2:

![cachemem](https://user-images.githubusercontent.com/105136494/171315398-6783037d-d447-458b-90a2-e57d5400e82a.png)

![Lenovo-s30](https://user-images.githubusercontent.com/105136494/171315454-afc9315f-cb38-46a2-bf77-689b7100d2cf.png)

As you can see, there is ~4x difference in memory benchmark, and about 10x(!) for L1/L2/L3 cache. For RAM, the speed is closer to what one would expect form a single DDR3 channel, where this this a quad channel setup.

(and yes, the SPD does not work, but same issue with CPU-Z so I would blame Lenovo here - not sure if this is bug worthy?)

Same issue for a single channel on a "am-I-joke-to-you" type of platform (Intel Baytrail-D Celeron Dual-Core J1800):

![cachemem](https://user-images.githubusercontent.com/105136494/171317263-8fef76b1-1e4f-4855-afab-572ce038b062.png)

![dell-inspiron-3646](https://user-images.githubusercontent.com/105136494/171317227-7abe2112-49e5-4ea4-99d4-ce098896dbdd.png)

(note that there is a single SO-DIMM socket on the MB, so AIDA is not correct saying Dual-Channel)

**Comment by Hypoon** [comment]
Part of this discussion seems to hinge on whether Memtest86+ needs to support 256+ threads. I want to raise a different motivation: for some processors/platforms, x2APIC (or similar) is required for even just 2+ threads. A BIOS change might relax this requirement, but I think it's preferable to let people optimize their BIOS for their primary OS, not for Memtest86+.

I can't comment (due to lack of expertise/insight) on whether Memtest86+ needs support for 2+ threads, but _if it doesn't_, then I imagine a lot of existing code could probably be removed, right?

**Comment by x86fr** [comment]
There is no debate about the need to support x2APIC (we need to). The debate is about the interest of running 128, 256 or even more thread at once because this can lower Memtest86+'s ability to detect memory issues in a timely manner. Spending much more time waiting for barriers or switching between threads instead of actually testing memory may reduce the test efficiency.

**Comment by Hypoon** [comment]
Ah, I misunderstood the discussion! Right now it takes me about 16 to 17 hours to test 384GB of memory. I would have thought that running 64 threads would speed that up significantly, but it depends. If the intent is to test every core's access to every bit of memory, then that could only slow things down. But if each core only needs to test 6GB of memory... surely that wouldn't take 16 to 17 hours anymore, right?

**Comment by debrouxl** [comment]
Testing every core's access to every bit of memory is the aim of tests 1 and 2, so these take approximately linear time with the number of cores and linear time with the amount of memory. However, the bulk of memtest86+'s operation is currently a matter of splitting the work on the same area of <= 1 GB across the enabled hardware threads.

x2APIC / AMD extended APIC support is needed for being able to use CPU cores with APIC IDs >= 256, but in practice, not for running 256 _simultaneous_ threads (especially on the same area of RAM), which, besides being a significant waste of resources, is probably detrimental to finding memory errors quickly and/or accurately, in the face of NUMA, which all recent high-end platforms are.
Since my first post last year, the rumored dual-socket EPYC Genoa 96C/192T server models have become available. They're definitely not an average computer, but they are closer to average than an uncommon, heavyweight and very expensive 8S server model.

Maximizing memory bandwidth usage, without wasting resources, is quite a bit of work, due to the diversity of CPU micro-architectures and their memory access characteristics, as well as the topology of the platforms the CPUs are powering - RAM slot population, etc.
I've recently started tackling an angle of that problem with initial NUMA awareness support, the #12 I mentioned above. I can't really recommend testing the experimental code yet, at least without disabling test 7 whose deadlock I'm still failing to fix. As I commented there, on NUMA computers, the NUMA-aware mode will usually consume (much) less energy and may save time.
Testing multiple memory ranges in parallel, one per NUMA node (proximity domain) or so, would be the next conceptual step for reducing the wall clock time taken by tests (with unspecified impact on memory error detection reliability), at the expense of resources. That requires changes in all of low-level code (a set of pml4, pdp, pd2 page tables per node, and the corresponding handling code), testing and support code (dealing with multiple simultaneous ranges), and certainly the UI code, for good measure. Not going to be done any time soon, if ever (assuming it is even deemed a path to follow in the first place).

**Comment by x86fr** [comment]
Support for x2APIC should really become our top priority as it's confirmed that Intel dropped default support for legacy APIC starting with Arrow Lake (released next month). Final platform can't do SMP testing on Memtest86+ due to the lack of x2APIC support. I dropped my try to re-enable legacy APIC. Anyone working on it already? Technically, adding x2APIC is not really complex, but debugging is quite long (messing with MSR lead to GPF Fiesta)...

**Comment by martinwhitaker** [comment]
My newest/most powerful CPU is a Zen 2 8C/16T, so no, I'm not working on this.
My top priority would be to revisit and apply the outstanding PRs, then reorganise the code properly for multiple architectures. But I haven't had enough free time/energy to tackle this yet.

**Comment by debrouxl** [comment]
I'm not working on this either. Nearly none of my computers which I can reboot at will is x2APIC-capable anyway.
I rebased several of my work branches this week, and I planned on looking at the architecture reorganization one of these days - to begin with, adding a `x86/` directory for x86/x86-64-specific files, the way it was done for `la64`.

**Comment by x86fr** [comment]
Intel officially announced the deprecation of xAPIC. 

https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/technical-documentation/xapic-deprecation-plan.html

I'm focusing on x2APIC support now to avoid losing SMP support for modern CPUs.

I have some working code that I will push later today. It deserves to be checked on a lot of computers, but I already tested it on a Ryzen 9 and a beefy EPYC server without issues. 

<img width="2802" height="1544" alt="Image" src="https://github.com/user-attachments/assets/4fa2e333-d803-446b-ae9a-038bb458591e" />

**Comment by debrouxl** [comment]
Sounds good :)
I can easily test on BDX-EP server and SKL laptop classes hardware; I have access to various generations of newer hardware, but can't do whatever I please with them.

At the same time we're adding support for x2APIC mode, shouldn't we try to flip the default on NUMA mode as well ? After all, many pieces of recent Intel and AMD hardware tend to grow intra-socket NUMA nodes again (Sub-NUMA Clustering, etc.), like Opteron Bulldozer and Piledriver had back in the day, due to their multi-chip (CCX, etc.) construction.

**Comment by x86fr** [comment]
> At the same time we're adding support for x2APIC mode, shouldn't we try to flip the default on NUMA mode as well ? After all, many pieces of recent Intel and AMD hardware tend to grow intra-socket NUMA nodes again (Sub-NUMA Clustering, etc.), like Opteron Bulldozer and Piledriver had back in the day, due to their multi-chip (CCX, etc.) construction.

Yes. I have a single report of very slow performance after enabling NUMA testing on a EPYC server (like 20x slower) but it's probably unrelated. I should have access to that server soon.

**Comment by debrouxl** [comment]
Yeah, such a factor of slowdown is odd...

I had noticed a slight slowdown of NUMA-aware mode on a 2S Westmere server configured in NUMA mode (Supermicro X8D family), but removing the inter-socket communication bottleneck is typically a large speed improvement. No idea about the impact on test reliability.
